agent/acs/handler/acs_handler.go

```diff
@@ -138,147 +102,128 @@ type sessionResources interface {
-	session := newSessionResources(args)
-	return startSession(ctx, args, backoff, session)
+// sessionState defines state recorder interface for the
+// session established with ACS. It can be used to record and
+// retrieve data shared across multiple connections to ACS
+type sessionState interface {
+	// connectedToACS callback indicates that the client has
+	// connected to ACS
+	connectedToACS()
+	// getSendCredentialsURLParameter retrieves the value for
+	// the 'sendCredentials' URL parameter
+	getSendCredentialsURLParameter() string
 }
 
-func newSessionResources(args StartSessionArguments) sessionResources {
-	return &acsSessionResources{
-		startSessionArguments: args,
-		sendCredentials:       true,
+// NewSession creates a new Session object
+func NewSession(ctx context.Context,
+	acceptInsecureCert bool,
+	config *config.Config,
+	deregisterInstanceEventStream *eventstream.EventStream,
+	containerInstanceArn string,
+	credentialsProvider *credentials.Credentials,
+	ecsClient api.ECSClient,
+	stateManager statemanager.StateManager,
+	taskEngine engine.TaskEngine,
+	credentialsManager rolecredentials.Manager) *Session {
+	resources := newSessionResources(config.AWSRegion, credentialsProvider, acceptInsecureCert)
+	backoff := utils.NewSimpleBackoff(connectionBackoffMin, connectionBackoffMax,
+		connectionBackoffJitter, connectionBackoffMultiplier)
+	derivedContext, cancel := context.WithCancel(ctx)
+
+	return &Session{
+		acceptInsecureCert:            acceptInsecureCert,
+		agentConfig:                   config,
+		deregisterInstanceEventStream: deregisterInstanceEventStream,
+		containerInstanceARN:          containerInstanceArn,
+		credentialsProvider:           credentialsProvider,
+		ecsClient:                     ecsClient,
+		stateManager:                  stateManager,
+		taskEngine:                    taskEngine,
+		credentialsManager:            credentialsManager,
+		ctx:                           derivedContext,
+		cancel:                        cancel,
+		backoff:                       backoff,
+		resources:                     resources,
 	}
 }
 
-// startSession creates a session with ACS and handles requests from ACS
-// It also tries to repeatedly connect to ACS when disconnected
-func startSession(ctx context.Context, args StartSessionArguments, backoff *utils.SimpleBackoff, acsResources sessionResources) error {
+// Start starts the session. It'll forever keep trying to connect to ACS unless:
+// 1. The context is cancelled
+// 2. Instance is deregistered
+//
+// If the context is cancelled, Start() would return with the error code returned
+// by the context.
+// If the instance is deregistered, Start() would emit an event to the
+// deregister-instance event stream and blocks until the ECS Agent is killed.
+func (s *Session) Start() error {
+	// connectToACS channel is used to inidcate the intent to connect to ACS
+	// It's processed by the select loop to connect to ACS
+	connectToACS := make(chan struct{})
```

Comment:
Making this a bool would mean that we attach some meaning to the message (if true, reconnect; if not don't), which is not really the case here. The type `chan struct{}` says that the channel contains no value; Any message on this channel means  that we reconnect to ACS. We use empty structs in the Agent at multiple modules to indicate the notion of "some work" that needs to be done when a message is available.